### PR TITLE
[API breaking] Use polymorphic value type holder for manifolds and constraint sets

### LINKED ISSUE #87: Template instantiation: split up `constraints.hpp` and `constraints.txx` mega-headers

### PR DESCRIPTION
**Pull request by ManifoldFR**
This PR:

- removes use of `shared_ptr` for manifold and constraint set types, switching to [jbcoe/value_types](https://github.com/jbcoe/value_types/)'s `polymorphic` template class everywhere as a holder with value semantics 
- enhances the library's ergonomics by avoiding having to call `std::make_shared<T>(...)` all the time, see changed C++ examples
- updates the examples and tests accordingly
- adds a util header `proxsuite-nlp/python/polymorphic.hpp` for dealing with values/references of `polymorphic<T>`, register conversions, etc
- updates Python bindings for manifolds and constraint sets accordingly
- adds a `PolymorphicVisitor` to the Python bindings which registers a type `Y` as convertible to the template argument `polymorphic<X>`
- splits up the big constraint instantiation header `constraints.txx`: closes #87 

### Known issues

- returning a `polymorphic<X>` by value does not check whether the underlying object is a `boost::python::wrapper<U>` for some type `U` which inherits from `X`
- **this PR is a breaking change for proxsuite-nlp's C++ API**

## Background

This is the first in a series of PR which will cut down on our overuse of `shared_ptr` across both proxsuite-nlp and [aligator](https://github.com/Simple-Robotics/proxsuite-nlp/pull/90) and replace it with a holder for polymorphic objects with value semantics.

The drawbacks to the current approach with `shared_ptr` was:
- no real copy constructor for `shared_ptr<T>`, except if we add virtual clone member functions implemented in all derived classes
- calls to members did not satisfy const-correctness (not a problem for manifolds as they so far have no non-const member functions, but a problem for namely `BoxConstraint`)
- **shared mutable state**

The polymorphic value type was described by Jonathan Coe: https://www.youtube.com/watch?v=sjLRX4WMvlU. It allows composite classes (with polymorphic data members) to behave well using value-semantics all the way down (with const-correctness).

### Why not unique_ptr ?

unique_ptr completely removes any ability to copy the underlying object (except if a virtual `clone()` was implemented) and disables default copy ctor/operators for all composite classes using it.

Many thanks to @edantec for his contribution.

### CODE REVIEW

**Comment by ManifoldFR** [comment]
The last thing to do before merging, which is _very_ relevant for downstream integration into aligator, is being able to _extend interfaces from Python_. Right now, inheriting from `bp::wrapper<PythonClass>` and using the current macros doesn't seem to work anymore. We need to:
- [ ] add this case to the test
- [ ] find a way to fix this. AFAIK, for `shared_ptr<T>`, Boost.Python handles calling virtual methods implemented in Python by fetching a pointer to the Python object that is hidden in the deleter

File: examples/ur5-ik.cpp [hunk 1]
```diff
@@ -90,18 +90,19 @@ int main() {
...
 
   constexpr bool has_joint_lims = true;
   if (has_joint_lims) {
-    auto box_cstr = std::make_shared<BoxConstraintTpl<Scalar>>(
-        model.lowerPositionLimit, model.upperPositionLimit);
+    BoxConstraintTpl<Scalar> box_cstr(model.lowerPositionLimit,
+                                      model.upperPositionLimit);
     ConstraintObjectTpl<Scalar> cstrobj(
-        std::make_shared<ManifoldDifferenceToPoint<Scalar>>(space, q0),
+        std::make_shared<ManifoldDifferenceToPoint<Scalar>>(PolyManifold{space},
```
Comment (edantec) [inline examples/ur5-ik.cpp:104]: Use of shared pointer here. Do we plan to take care of it in this PR or the next?

File: examples/ur5-ik.cpp [hunk 2]
```diff
@@ -90,18 +90,19 @@ int main() {
       allocate_shared_eigen_aligned<QuadraticResidualCostTpl<Scalar>>(fn, w1);
   auto cost2 = allocate_shared_eigen_aligned<QuadraticDistanceCostTpl<Scalar>>(
       space, q0, w2);
-  auto cost = std::make_shared<CostSumTpl<Scalar>>(space->nx(), space->ndx());
+  auto cost = std::make_shared<CostSumTpl<Scalar>>(space.nx(), space.ndx());
```
Comment (edantec) [inline examples/ur5-ik.cpp:93]: Use of shared pointer here. Do we plan to take care of it in this PR or the next?

File: include/proxsuite-nlp/constraint-base.hpp [hunk 3]
```diff
@@ -125,7 +124,7 @@ template <typename _Scalar> struct ConstraintObjectTpl {
   using ConstraintSet = ConstraintSetBase<Scalar>;
 
   shared_ptr<FunctionType> func_;
```
Comment (edantec) [inline include/proxsuite-nlp/constraint-base.hpp:126]: Same comment about shared ptr

**Review by edantec — COMMENTED** [review]
After quick glance, PR seems ok

File: examples/ur5-ik.cpp (hunk 1 above)
Comment (ManifoldFR) [reply examples/ur5-ik.cpp:104]: For cost functions and residuals, we plan to do this in a later PR. For now we are doing this for the components of proxsuite-nlp which are used downstream in aligator.

**Comment by ManifoldFR** [comment]
@jorisv @edantec I think Windows is finally fixed (my god...)

File: bindings/python/include/proxsuite-nlp/python/polymorphic.hpp [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+  using Poly = xyz::polymorphic<Base, A>;
+  static_assert(std::is_polymorphic_v<Base>, "Type should be polymorphic!");
+
+  template <class PyClass> void visit(PyClass &) const {
+    using T = typename PyClass::wrapped_type;
+    using meta = typename PyClass::metadata;
+    using held = typename meta::held_type;
+    typedef bp::converter::implicit<held, Poly> functions;
+
+#pragma GCC diagnostic push
```
Comment (jorisv) [inline bindings/python/include/proxsuite-nlp/python/polymorphic.hpp:74]: Do something similar than pinocchio to manage push/pop and warnings with multiple compiler: https://github.com/stack-of-tasks/pinocchio/blob/devel/include/pinocchio/macros.hpp#L99-L130

**Review by jorisv — CHANGES_REQUESTED** [review]

File: bindings/python/include/proxsuite-nlp/python/polymorphic.hpp (hunk 4 above)
Comment (ManifoldFR) [reply bindings/python/include/proxsuite-nlp/python/polymorphic.hpp:74]: wow, that is rather complex code

**Comment by ManifoldFR** [comment]
While discussing with @jorisv, we found that:

- passing the wrapper class to a function/ctor taking `polymorphic<T>` does not properly copy the object
- by properly copying, we mean that the reference to the initial object is not properly handled, there's a missing incref somewhere
- in Boost.Python, the converter class `shared_ptr_from_python`, which is registered in the `class_metadata.hpp` header, takes the source `PyObject*`, borrows it with `bp::borrowed` (which casts it to a pointer to the special type `bp::detail::borrowed<PyObject>` (see `boost/python/detail/borrowed_ptr.hpp`)), and shoves it into a `bp::handle<>` object, the constructor of which increfs the pointer to `borrowed` and upcasts it back to `PyObject` to store it.

**Comment by edantec** [comment]
Are we ready to merge this yet?

**Review by jorisv — APPROVED** [review]

**Comment by ManifoldFR** [comment]
Okay, we will be merging this today. This is not the perfect solution imo, I'd be willing to look at [other types of type erasure](https://akrzemi1.wordpress.com/2013/12/06/type-erasure-part-ii/) in the future for the definitive API (and rip out the OOP class hierarchies here and in aligator, downstream).
Having value semantics is already good enough though and we can make progress on other fronts.

**Pull request merged**